mono/metadata/oop.c

```diff
@@ -0,0 +1,422 @@
+/*
+ * oop.c: These functions allow us to access the MonoDomain internals for purposes of post-mortem
+ * inspection by another process. All data is immutable: these calls are guaranteed to have no 
+ * side-effects. These routines are not thread safe. This does not work with AOT modules.
+ *
+ * Author:
+ *	Pete Lewis <pete.lewis@unity3d.com>
+ *
+ * Copyright 2017 Unity Technologies (http://www.unity3d.com)
+ * Copyright 2001-2003 Ximian, Inc (http://www.ximian.com)
+ * Copyright 2004-2009 Novell, Inc (http://www.novell.com)
+ */
+
+#include <config.h>
+#include <glib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#include <mono/utils/mono-compiler.h>
+#include <mono/utils/mono-logger.h>
+#include <mono/utils/mono-membar.h>
+#include <mono/utils/mono-counters.h>
+#include <mono/metadata/object.h>
+#include <mono/metadata/object-internals.h>
+#include <mono/metadata/domain-internals.h>
+#include <mono/metadata/class-internals.h>
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/exception.h>
+#include <mono/metadata/metadata-internals.h>
+#include <mono/metadata/gc-internals.h>
+#include <mono/metadata/appdomain.h>
+#include <mono/metadata/mono-debug-debugger.h>
+#include <mono/metadata/mono-config.h>
+#include <mono/metadata/threads-types.h>
+#include <metadata/threads.h>
+#include <metadata/profiler-private.h>
+#include <mono/metadata/coree.h>
+
+#ifdef _M_X64
+#include <mono/mini/mini-amd64.h>
+extern GList* g_dynamic_function_table_begin;
+extern SRWLOCK g_dynamic_function_table_lock;
+#endif
+
+// petele: todo: move this structure into a mono header
+typedef struct _MonoStackFrameDetails
+{
+    char* methodName;
+    size_t methodNameLen;
+    char* className;
+    size_t classNameLen;
+    char* assemblyName;
+    size_t assemblyNameLen;
+} MonoStackFrameDetails;
+
+typedef gboolean(*ReadMemoryCallback)(void* buffer, gsize* read, const void* address, gsize size, void* userdata);
+typedef gboolean(*ReadExceptionCallback)(const void* address, gsize size, void* userdata);
+
+typedef struct _OutOfProcessMono
+{
+    ReadMemoryCallback readMemory;
+    ReadExceptionCallback readException;
+    void* userData;
+} OutOfProcessMono;
+
+static OutOfProcessMono g_oop = { NULL, NULL };
+
+#define OFFSET_MEMBER(type, base, member) ((gpointer)((gchar*)(base) + offsetof(type, member)))
+
+void read_exception(const void* address, gsize size)
+{
+    if (g_oop.readException)
+        g_oop.readException(address, size, g_oop.userData);
+    else
+        abort();
```

Comment:
I think this should be `g_assert_not_reached` instead of `abort`, as the former will give us a stack trace and some diagnostic information if we reach this line of code. I don't think `abort` will do that on some platforms.